statement_timeout parameter dose not cancel Babelfish session

aadityavermaa opened the pull request:
### Description

statement_timeout parameter aborts any statement that takes more than the specified amount of time, currently it does not kill Babelfish session when it exceeds the timeout. We have added enable_statement_timeout and disable_statement_timeout to fix this.

Task: BABEL-1914

[List any issues this PR will resolve]

### Test Scenarios Covered ###
* **Use case based -**
- tests with query execution time more than statement_timeout parameter value
- tests with query execution time less than statement_timeout parameter value
- test behaviour when statement_timeout is disabled and when it is equal to guc value 

* **Boundary conditions -**


* **Arbitrary inputs -**


* **Negative test cases -**


* **Minor version upgrade tests -**


* **Major version upgrade tests -**


* **Performance tests -**


* **Tooling impact -**


* **Client tests -**



### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
BABEL-1914:statement_timeout param doesn't cancel Bbf session

- added enable_statement_timeout and disable_statement_timeout in
  tdsprotocol.c

Signed-off-by: Aditya Verma <adiityav@amazon.com>

Commit:
BABEL-1914: statement_timeout doesn't cancel bbf session

-- added comments related to changes and testcases for the
implementation

Signed-off-by: Aditya Verma <adiityav@amazon.com>

KushaalShroff commented on contrib/babelfishpg_tds/src/backend/tds/tdsprotocol.c:
```diff
@@ -398,6 +406,10 @@ ProcessTDSRequest(TDSRequest request)
 				break;
 		}
+
+		/* Disabling statement timeout after processing */
```
Here the comment should end with full stop.
if the sentence starts with a small case letter then there is no need of a full stop

KushaalShroff commented on contrib/babelfishpg_tds/src/backend/tds/tdsprotocol.c:
```diff
@@ -289,6 +293,10 @@ GetTDSRequest(bool *resetProtocol)
 			return NULL;
 		}
 
+		/* Enable statement timeout. Note we add this function here to
+		 * include the time taken by the protocol in the timeout  */
```
In Postgres we follow the following patter for a multi-line comment:
```
/*
 * Enable statement timeout. Note we add this function here to
 * include the time taken by the protocol in the timeout.
 * /
```
Also it should end with full-stop

Commit:
BABEL-1914:

- addressed review comments

Signed-off-by: Aditya Verma <adiityav@amazon.com>

surendravishnoi commented on contrib/babelfishpg_tds/src/backend/tds/tdsprotocol.c:
```diff
@@ -711,3 +725,31 @@ TestGetTdsRequest(uint8_t reqType, const char *expectedStr)
+{
+	if (StatementTimeout > 0)
+	{
+		if (!get_timeout_active(STATEMENT_TIMEOUT))
```
what does this mean? if timer is already running do not do anything, shouldn't we reset the timeout to guc value

aadityavermaa replied:
Why is it needed to reset the guc value ?

surendravishnoi replied:
not to reset guc value but reset timer to guc value. We want to wait for statement_timeout, if timer is already running then we will only wait for remaining time at this point in time

surendravishnoi commented on contrib/babelfishpg_tds/src/backend/tds/tdsprotocol.c:
```diff
@@ -398,6 +408,10 @@ ProcessTDSRequest(TDSRequest request)
 		}
+
+		/* Disabling statement timeout after processing. */
+		disable_statement_timeout();
```
how are we making sure that a running timer is not leaked due error conditions etc?

kuntalghosh replied:
We should call this function as the first statement in PG_CATCH block as well.

surendravishnoi reviewed (commented):
Tests are missing for boundary conditions (min timeout, 0 etc), negative cases and arbitrary input

Commit:
BABEL-1914 - statemet_timeout doesn't cancel bbf session

- addressed review comments
	- fixed code comments
	- added disable_statement_timeout in PG_CATCH
	- updated test case file

Signed-off-by: Aditya Verma <adiityav@amazon.com>

jsudrik commented:
statement_timeout parameter aborts any statement that takes more than the specified amount of time, currently it does not kill Babelfish session when it exceeds the timeout. We have added enable_statement_timeout and disable_statement_timeout to fix this.
